src/threads_arg/normalization.py

```diff
@@ -0,0 +1,122 @@
+import pandas
+from threads_arg import ThreadingInstructions
+import msprime
+
+logging.basicConfig(
+    format='%(asctime)s %(levelname)-8s %(message)s',
+    level=logging.INFO,
+    datefmt='%Y-%m-%d %H:%M:%S')
+logging.getLogger("msprime").setLevel(logging.WARNING)
+
+class Normalizer:
+    """
+    Minimal simulator based on ARG-Needle Simulator class
+    and including the ARG-Needle normalisation routine.
+    """
+    def __init__(self, demography_file, num_samples):
+        self.num_samples = num_samples
+        self.demography_file = demography_file
+        self.demography = self.read_demography(demography_file)
+
+    def read_demography(self, demography_file):
+        df = pd.read_table(demography_file, header=None)
+        df.columns  = ['GEN', 'NE']
+        demography = msprime.Demography()
+        # NOTE: these initial sizes get overwritten anyways
+        demography.add_population(name="A", initial_size=1e4)
+        for t, ne in zip(df["GEN"], df["NE"]):
+            # Divide by 2 because msprime wants diploids but demography is haploid
+            demography.add_population_parameters_change(t, initial_size=ne / 2, population="A")
+        return demography
+
+    def simulation(self, length, random_seed=10):
+        ts = msprime.sim_ancestry(
+            samples={"A": self.num_samples // 2},  # again, diploid
+            demography=self.demography,
+            sequence_length=length,
+            recombination_rate=0,
+            random_seed=random_seed)
+        return ts
+
+    def normalize(self, threading_instructions, num_seeds=1000, start_seed=1):
+        num_samples = threading_instructions.num_samples
+        assert num_samples == self.num_samples
+        
+        heights_to_span = {}
+        positions = threading_instructions.positions
+
+        for start_vec, tmrca_vec in zip(threading_instructions.all_starts(), threading_instructions.all_tmrcas()):
+            assert len(start_vec) == len(tmrca_vec)
+            for i, (start, tmrca) in enumerate(zip(start_vec, tmrca_vec)):
+                try:
+                    span = start_vec[i + 1] - start
+                except IndexError:
+                    span = positions[-1] - start
+                heights_to_span[tmrca] = heights_to_span[tmrca] + span if tmrca in heights_to_span else span
+
+        numpy_stuff = np.array(sorted(heights_to_span.items()))
+
+        cumsum = np.cumsum(numpy_stuff[:, 1])
+        quantiles = (cumsum - 0.5 * numpy_stuff[:, 1]) / cumsum[-1]
+
+        node_times = []
+
+        for seed_offset in range(num_seeds):
+            seed = seed_offset + start_seed
+
+            # 1e6 doesn't matter here as rho = 0
+            simulation = self.simulation(1e6, random_seed=seed)
+            highest = 0
+            for node in simulation.nodes():
+                if node.time > 0:
+                    node_times.append(node.time)
+                    highest = max(node.time, highest)
+
+        node_times.sort()
+
```

Comment:
A brief one liner comment on logic of correction_dict generation (e.g. use of sim_quantiles, node_times_pad, how that relates to new tmrcas), would help break these chunks of code up and orient the reader.

Optional: It looks like each of the chunks of code in this function has a particular responsibility (e.g. generating heights_to_span lookup, generate node_times), so similarly a brief one-liner comment may help future readers.